pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/AdditionTransformFunction.java

```diff
@@ -48,14 +49,33 @@ public void init(@Nonnull List<TransformFunction> arguments, @Nonnull Map<String
       throw new IllegalArgumentException("At least 2 arguments are required for ADD transform function");
     }
 
-    for (TransformFunction argument : arguments) {
-      if (argument instanceof LiteralTransformFunction) {
-        _literalSum += Double.parseDouble(((LiteralTransformFunction) argument).getLiteral());
+    checkOperands(arguments);
+  }
+
+  private void checkOperands(List<TransformFunction> operands) {
+    for (TransformFunction operand : operands) {
+      if (operand instanceof MapValueTransformFunction) {
+        throw new IllegalArgumentException("ADD transform function not supported to work with MAP as inner transform function");
+      }
+
+      if (operand instanceof LiteralTransformFunction) {
+        try {
+          _literalSum += Double.parseDouble(((LiteralTransformFunction) operand).getLiteral());
+        } catch (NumberFormatException ne) {
+          throw new IllegalArgumentException("ADD transform function not supported on non-numeric literals");
+        }
       } else {
-        if (!argument.getResultMetadata().isSingleValue()) {
-          throw new IllegalArgumentException("All the arguments of ADD transform function must be single-valued");
+        final TransformResultMetadata resultMetadata = operand.getResultMetadata();
+
+        if (resultMetadata.getDataType() == FieldSpec.DataType.STRING) {
+          throw new IllegalArgumentException("ADD transform function not supported on non-numeric types");
```

Comment:
Let's change all IllegalArgumentException to IllegalStateException as the arguments are not provided by the client

Reply:
done

Reply:
??? Still IllegalArgumentException everywhere

Reply:
You can use `Preconditions.checkState()` for simplicity